src/workloads/execution/FilterWithLogicalExpressions.yml

```diff
@@ -0,0 +1,162 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/query"
+Description: >
+  This workload stresses the query execution engine by running queries with complex logical
+  expressions that never match a document in the collection.
+
+GlobalDefaults:
+  Database: &Database test
+  Collection: &Collection Collection0
+  DocumentCount: &DocumentCount 1e6
+  MaxPhases: &MaxPhases 5
+
+Actors:
+# Phase 0: Insert documents into the collection.
+- Name: InsertData
+  Type: Loader
+  Threads: 1
+  Phases:
+    OnlyActiveInPhases:
+      Active: [0]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: 1
+        Database: *Database
+        Threads: 1
+        CollectionCount: 1
+        DocumentCount: *DocumentCount
+        BatchSize: 1000
+        Document:
+          # NOTE: These documents will not produce a match for the filter below, forcing the query
+          # to scan the entire collection.
+          a: 1
+          b: 1
+          c: 1
+          d: 1
+
+# Phase 1: Ensure all data is synced to disk.
+- Name: Quiesce
+  Type: QuiesceActor
+  Threads: 1
+  Database: *Database
+  Phases:
+    OnlyActiveInPhases:
+      Active: [1]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: 1
+        Threads: 1
+
+# Phase 2: Run query with wide expression.
+- Name: RunWideQuery
+  Type: CrudActor
+  Database: *Database
+  Threads: 1
+  Phases:
+    OnlyActiveInPhases:
+      Active: [2]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: 10
+        Collection: *Collection
+        Operations:
+        - OperationName: find
+          OperationCommand:
+            Filter: {
+              $expr: {
+                $or: [
+                  {$eq: ["$a", 10]},
+                  {$eq: ["$b", 11]},
+                  {$eq: ["$c", 12]},
+                  {$eq: ["$d", 13]},
+                  {$eq: ["$a", 14]},
+                  {$eq: ["$b", 15]},
+                  {$eq: ["$c", 16]},
+                  {$eq: ["$d", 17]},
+                  {$eq: ["$a", 18]},
+                  {$eq: ["$b", 19]},
```

Comment:
I see that these only test equality. I _think_ that's ok, since it would be overkill to try to have complete test matrix where we have wide/deep logical expressions involving inequalities or any other leaf expression we can think of. Presumably other expressions would benefit from a similar optimization strategy and we will implement these optimizations in a general way that would apply to $lt, $gt, etc. But we just won't be able to reasonably add perf coverage for such cases explicitly. Does this match your understanding? Are there any particularly important expressions other than `$eq` that you think we should benchmark?

Reply:
Replied in Slack.

Reply:
The Slack conversation suggests that we are on the same page. Do you want to copy any of the context here for posterity before resolving this thread?

Reply:
In the end, we decided to limit the benchmarks to the $eq expressions and derive some more optimizations for the type checker from analyzing common patterns in the stage builders.